parlai/tasks/airdialogue/agents.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from parlai.core.teachers import FixedDialogTeacher
+from .build import build
+import os
+import json
+
+
+def _path(opt):
+    # build the data if it does not exist
+    build(opt)
+
+    # set up path to data (specific to each dataset)
+    jsons_path = os.path.join(
+        opt['datapath'], 'airdialogue', 'airdialogue_data', 'airdialogue'
+    )
+    return jsons_path
+
+
+class AirDialogueTeacher(FixedDialogTeacher):
+    """
+    AirDialogue Teacher.
+
+    This also contains other files (dev_kb.json, train_kb.json) with flight data about
+    return flights, price, connections, flight airlines, departure times, and other
+    flight information.
+    """
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        jsons_path = _path(opt)
+        self.datatype = opt['datatype'].split(':')[0]
+        self.messages = []
+        self._setup_data(jsons_path)
+        self.id = 'airdialogue'
+        self.reset()
+
+    def _setup_data(self, jsons_path):
+        train_path = os.path.join(jsons_path, 'train_data.json')
+        test_valid_path = os.path.join(jsons_path, 'dev_data.json')
+        if self.datatype.startswith('test'):
+            with open(test_valid_path) as f:
```

Comment:
looks like you have very similar blocks of code for test/valid/train. could we try making htis a bit more modular by calling a helper function?